File: RenderLab/Core/CoreInterpolationBridge.cpp
//
//  CoreInterpolationBridge.cpp
//  RenderLab
//
//  C bridge wrappers for Interpolation Lab math and uniform computation.
//

#include "CoreBridge.h"
#include "CoreInterpolation.h"

namespace {
CoreInterpScalarMode scalarModeFromRaw(int32_t raw) {
    switch (raw) {
        case CORE_INTERP_SCALAR_SMOOTHSTEP:
            return CORE_INTERP_SCALAR_SMOOTHSTEP;
        case CORE_INTERP_SCALAR_CUBIC:
            return CORE_INTERP_SCALAR_CUBIC;
        case CORE_INTERP_SCALAR_LERP:
        default:
            return CORE_INTERP_SCALAR_LERP;
    }
}

CoreInterpRotationMode rotationModeFromRaw(int32_t raw) {
    switch (raw) {
        case CORE_INTERP_ROT_EULER_LERP:
            return CORE_INTERP_ROT_EULER_LERP;
        case CORE_INTERP_ROT_QUAT_NLERP:
            return CORE_INTERP_ROT_QUAT_NLERP;
        case CORE_INTERP_ROT_QUAT_SLERP:
        default:
            return CORE_INTERP_ROT_QUAT_SLERP;
    }
}

CoreInterpLoopMode loopModeFromRaw(int32_t raw) {
    switch (raw) {
        case CORE_INTERP_LOOP_REPEAT:
            return CORE_INTERP_LOOP_REPEAT;
        case CORE_INTERP_LOOP_PINGPONG:
            return CORE_INTERP_LOOP_PINGPONG;
        case CORE_INTERP_LOOP_CLAMP:
        default:
            return CORE_INTERP_LOOP_CLAMP;
    }
}

CoreTransform toCoreTransform(const CoreSceneTransform& in) {
    CoreTransform out{};
    out.position = {in.position[0], in.position[1], in.position[2]};
    out.rotation = {in.rotation[0], in.rotation[1], in.rotation[2]};
    out.scale = {in.scale[0], in.scale[1], in.scale[2]};
    return out;
}

CoreSceneTransform toBridgeTransform(const CoreTransform& in) {
    CoreSceneTransform out{};
    out.position[0] = in.position.x;
    out.position[1] = in.position.y;
    out.position[2] = in.position.z;
    out.rotation[0] = in.rotation.x;
    out.rotation[1] = in.rotation.y;
    out.rotation[2] = in.rotation.z;
    out.scale[0] = in.scale.x;
    out.scale[1] = in.scale.y;
    out.scale[2] = in.scale.z;
    return out;
}

CoreInterpConfigCpp toCoreConfig(const CoreInterpConfig* config) {
    const CoreInterpConfigCpp defaults = CoreInterpMakeDefaultConfig();
    if (!config) return defaults;

    CoreInterpConfigCpp out{};
    out.positionMode = scalarModeFromRaw(config->positionMode);
    out.rotationMode = rotationModeFromRaw(config->rotationMode);
    out.scaleMode = scalarModeFromRaw(config->scaleMode);
    out.shortestPath = config->shortestPath != 0;
    return out;
}

CoreInterpPlaybackStateCpp toCorePlaybackState(const CoreInterpPlaybackState* state) {
    const CoreInterpPlaybackStateCpp defaults = CoreInterpMakeDefaultPlaybackState();
    if (!state) return defaults;

    CoreInterpPlaybackStateCpp out{};
    out.t = state->t;
    out.speed = state->speed;
    out.loopMode = loopModeFromRaw(state->loopMode);
    out.isPlaying = state->isPlaying != 0;
    out.direction = state->direction;
    return out;
}

CoreInterpPlaybackState toBridgePlaybackState(const CoreInterpPlaybackStateCpp& state) {
    CoreInterpPlaybackState out{};
    out.t = state.t;
    out.speed = state.speed;
    out.loopMode = state.loopMode;
    out.isPlaying = state.isPlaying ? 1u : 0u;
    out.direction = state.direction;
    return out;
}
} // namespace

void coreInterpSetDefaultConfig(CoreInterpConfig* outConfig) {
    if (!outConfig) return;
    const CoreInterpConfigCpp defaults = CoreInterpMakeDefaultConfig();
    outConfig->positionMode = defaults.positionMode;
    outConfig->rotationMode = defaults.rotationMode;
    outConfig->scaleMode = defaults.scaleMode;
    outConfig->shortestPath = defaults.shortestPath ? 1u : 0u;
}

void coreInterpSetDefaultPlaybackState(CoreInterpPlaybackState* outState) {
    if (!outState) return;
    const CoreInterpPlaybackStateCpp defaults = CoreInterpMakeDefaultPlaybackState();
    *outState = toBridgePlaybackState(defaults);
}

int32_t coreInterpAdvancePlaybackState(CoreInterpPlaybackState* ioState,
                                       float deltaSeconds,
                                       float* outT) {
    if (!ioState) return 0;

    CoreInterpPlaybackStateCpp state = toCorePlaybackState(ioState);
    const bool changed = CoreInterpAdvancePlaybackState(state, deltaSeconds);
    *ioState = toBridgePlaybackState(state);
    if (outT) *outT = ioState->t;
    return changed ? 1 : 0;
}

int32_t coreInterpEvaluateTransform(const CoreSceneTransform* a,
                                    const CoreSceneTransform* b,
                                    float t,
                                    const CoreInterpConfig* config,
                                    CoreSceneTransform* outTransform,
                                    CoreInterpDebug* outDebug) {
    if (!a || !b || !outTransform) return 0;

    const CoreTransform transformA = toCoreTransform(*a);
    const CoreTransform transformB = toCoreTransform(*b);
    const CoreInterpConfigCpp coreConfig = toCoreConfig(config);

    CoreTransform outCoreTransform{};
    CoreInterpDebugCpp debug{};
    const bool ok = CoreInterpEvaluateTransform(
        transformA,
        transformB,
        t,
        coreConfig,
        outCoreTransform,
        outDebug ? &debug : nullptr
    );
    if (!ok) return 0;

    *outTransform = toBridgeTransform(outCoreTransform);
    if (outDebug) {
        outDebug->alphaPosition = debug.alphaPosition;
        outDebug->alphaScale = debug.alphaScale;
        outDebug->distanceToA = debug.distanceToA;
        outDebug->distanceToB = debug.distanceToB;
    }
    return 1;
}

int32_t coreInterpMakeObjectUniforms(const CoreUniforms* baseUniforms,
                                     const CoreSceneTransform* a,
                                     const CoreSceneTransform* b,
                                     float t,
                                     const CoreInterpConfig* config,
                                     CoreUniforms* outUniforms,
                                     CoreSceneTransform* outTransform,
                                     CoreInterpDebug* outDebug) {
    if (!baseUniforms || !a || !b || !outUniforms) return 0;

    CoreSceneTransform evaluated{};
    CoreInterpDebug debug{};
    const int32_t ok = coreInterpEvaluateTransform(
        a,
        b,
        t,
        config,
        &evaluated,
        outDebug ? &debug : nullptr
    );
    if (ok == 0) return 0;

    coreSceneMakeObjectUniforms(outUniforms, baseUniforms, &evaluated);
    if (outTransform) {
        *outTransform = evaluated;
    }
    if (outDebug) {
        *outDebug = debug;
    }
    return 1;
}

int32_t coreInterpMakeGhostUniforms(const CoreUniforms* baseUniforms,
                                    const CoreSceneTransform* a,
                                    const CoreSceneTransform* b,
                                    CoreUniforms* outUniformsA,
                                    CoreUniforms* outUniformsB) {
    if (!baseUniforms || !a || !b || !outUniformsA || !outUniformsB) return 0;

    coreSceneMakeObjectUniforms(outUniformsA, baseUniforms, a);
    coreSceneMakeObjectUniforms(outUniformsB, baseUniforms, b);
    return 1;
}

File: RenderLab/App/Renderer/Bridge/CoreInterpolationBridge.swift
//
//  CoreInterpolationBridge.swift
//  RenderLab
//
//  Swift wrappers over Interpolation Lab C bridge APIs.
//

import Foundation

enum CoreInterpolationBridge {
    static func defaultConfig() -> CoreInterpConfig {
        var config = CoreInterpConfig()
        coreInterpSetDefaultConfig(&config)
        return config
    }

    static func defaultPlaybackState() -> CoreInterpPlaybackState {
        var state = CoreInterpPlaybackState()
        coreInterpSetDefaultPlaybackState(&state)
        return state
    }

    @discardableResult
    static func advancePlayback(
        _ state: inout CoreInterpPlaybackState,
        deltaSeconds: Float
    ) -> Bool {
        var t: Float = state.t
        let changed = coreInterpAdvancePlaybackState(&state, deltaSeconds, &t)
        state.t = t
        return changed != 0
    }

    static func evaluateTransform(
        a: SceneTransform,
        b: SceneTransform,
        t: Float,
        config: CoreInterpConfig
    ) -> (transform: SceneTransform, debug: CoreInterpDebug)? {
        var transformA = a.toCoreSceneTransform()
        var transformB = b.toCoreSceneTransform()
        var config = config
        var outTransform = CoreSceneTransform()
        var outDebug = CoreInterpDebug()
        let ok = coreInterpEvaluateTransform(
            &transformA,
            &transformB,
            t,
            &config,
            &outTransform,
            &outDebug
        )
        guard ok != 0 else { return nil }
        return (SceneTransform.fromCoreSceneTransform(outTransform), outDebug)
    }

    static func makeObjectUniforms(
        baseUniforms: CoreUniforms,
        a: SceneTransform,
        b: SceneTransform,
        t: Float,
        config: CoreInterpConfig
    ) -> (uniforms: CoreUniforms, transform: SceneTransform, debug: CoreInterpDebug)? {
        var baseUniforms = baseUniforms
        var transformA = a.toCoreSceneTransform()
        var transformB = b.toCoreSceneTransform()
        var config = config
        var outUniforms = CoreUniforms()
        var outTransform = CoreSceneTransform()
        var outDebug = CoreInterpDebug()

        let ok = coreInterpMakeObjectUniforms(
            &baseUniforms,
            &transformA,
            &transformB,
            t,
            &config,
            &outUniforms,
            &outTransform,
            &outDebug
        )
        guard ok != 0 else { return nil }
        return (outUniforms, SceneTransform.fromCoreSceneTransform(outTransform), outDebug)
    }

    static func makeGhostUniforms(
        baseUniforms: CoreUniforms,
        a: SceneTransform,
        b: SceneTransform
    ) -> (uniformsA: CoreUniforms, uniformsB: CoreUniforms)? {
        var baseUniforms = baseUniforms
        var transformA = a.toCoreSceneTransform()
        var transformB = b.toCoreSceneTransform()
        var outUniformsA = CoreUniforms()
        var outUniformsB = CoreUniforms()

        let ok = coreInterpMakeGhostUniforms(
            &baseUniforms,
            &transformA,
            &transformB,
            &outUniformsA,
            &outUniformsB
        )
        guard ok != 0 else { return nil }
        return (outUniformsA, outUniformsB)
    }
}
